Add Ruby Gem wrapper / CLI interface and CI setup

iangmaia opened the pull request:
Related: [AINFRA-724: Wrap Rust CLI tool into a Gem and use it instead of a `release-toolkit` action](https://linear.app/a8c/issue/AINFRA-724/wrap-rust-cli-tool-into-a-gem-and-use-it-instead-of-a-release-toolkit)

This PR completes the `fluent-tools` tool by adding a Ruby gem wrapper and command-line interface for the Rust converter functionality and the CI setup to publish it.

### What's included:
- **Ruby gem structure**: Complete gemspec with proper dependencies and configuration
- **Ruby wrapper & CLI**: Bridge between Ruby and the underlying Rust converter
- **Ruby Installation script**: Automated setup for development and deployment
- **Initial CI setup**: Buildkite pipeline and scripts to publish the Gem and Rust binaries

### Dependencies:
- This PR is based on #2 and requires it to be merged first
- Utilizes both PO and Android XML conversion capabilities from previous PRs

This completes the initial implementation, providing multiple ways to access the Fluent conversion functionality: direct Rust usage, Ruby gem, and command-line interface.

iangmaia committed:
Add Ruby gem wrapper and CLI interface

iangmaia committed:
Add project Makefile

iangmaia committed:
Fix path finding in installer.rb

iangmaia committed:
Refactor Rakefile to better support different build types (native vs via docker)

iangmaia committed:
Add script to push the gem to RubyGems

iangmaia committed:
Add script to build and upload Rust tool binary

iangmaia committed:
Add main Buildkite pipeline file

iangmaia committed:
Improve Ruby Rubocop setup

iangmaia committed:
Add shared pipeline vars and IMAGE_ID reference in main pipeline

iangmaia committed:
Update code to install Rust for building natively on mac and building via Docker on mac agents

iangmaia committed:
Update code to build natively on Linux agents

iangmaia committed:
Fix Rakefile log message about release upload

iangmaia committed:
Remove unnecessary gemspec post install message

iangmaia committed:
Use 1 exit code when the Rust CLI is misused

iangmaia committed:
Reorganize and make Rakefile more consistent

iangmaia committed:
Improve Gem extension installation

iangmaia committed:
Fix Rubocop offenses and simplify find_binary_path

AliSoftware commented on .buildkite/pipeline.yml:
```diff
@@ -0,0 +1,81 @@
+      - tests
+      - rubocop
+    steps:
+      - label: ":linux: Release Binary"
+        command: |
+          echo "--- :rust: Installing Rust"
+          curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -v -y
+          source "$$HOME/.cargo/env"
+
+          .buildkite/release-binary.sh native
+        plugins: [$CI_TOOLKIT]
+        agents:
+          queue: android
+
+      - label: ":apple: Release Binary"
+        command: |
+          echo "--- :rust: Installing Rust"
+          curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -v -y
+          source "$$HOME/.cargo/env"
+
+          .buildkite/release-binary.sh native
+        plugins: [$CI_TOOLKIT]
+        agents:
+          queue: mac
```
Can't we run these in a `Rust` docker container, instead of installing `rustup` directly inside the CI agents?

Especially installing it directly inside an `android` agent / AWS EC2 instance that might be re-used by other CI jobs after it etc… I'd prefer us not to use `rustup` directly on the agents (for `mac` that's less of an issue since those already run in throw-away VMs… though if we had an already-prep'd Docker image instead of having to install `rustup` and Rust in it, that'd still be better isolation IMHO than having to install it ourselves?)

iangmaia replied:
Yeah, using Docker was my initial idea. I think one issue I had was with having Ruby in the `default` queue...
For inspiration, I checked how `wordpress-rs` does it and we do install ([1](https://github.com/Automattic/wordpress-rs/blob/0472bf8d263f21b11e4f8e45d426256b42267908/.buildkite/pipeline.yml#L63), [2](https://github.com/Automattic/wordpress-rs/blob/0472bf8d263f21b11e4f8e45d426256b42267908/.buildkite/pipeline.yml#L164)) `rustup` as well and it didn't sound that bad (as the install is pretty fast, `android` instances are recycled and the `mac` VM image is short lived).

But indeed, I'd prefer have the nice isolation that you can replicate in any environment that Docker provides. I think I'll give it another try, but I guess it also depends if the image supports different architecture targets 🤔

iangmaia replied:
I've iterated on the build setup on c0dbdc43730dc6964a26106ba3046c8b7b6cc14f.

I've managed to get the project compilation working via Docker on Linux 🎉  On macOS, the Docker cross-compilation linking would always fail. For now I kept it how it was (installing `rustup`), but now with the addition of being able to specify different architectures.

AliSoftware commented on ruby/Rakefile:
Having _both_ a `Rakefile` (here) and a `Makefile` (top level) is kinda confusing and mixed tech 🤔

iangmaia replied:
I can see that -- though I see them as having different purposes here: the `Makefile` is for building the Rust part, running Rust tests, running build commands with Docker, etc.
The `Rakefile` is focused on building and publishing the Gem (and it does other things that'd be overly complex in the Makefile like uploading binaries to GitHub).

I think the best alternative to the Rakefile would be Fastlane in this case? I think considered while I was trying things out but I thought it would be an overkill, but now taking a step back it doesn't sound bad 🤔

AliSoftware replied:
Ah right, I see the distinction now 👍  (that's how confusing reviewing a diff flat in GitHub can be compared to seeing it in an IDE where folder structures are more apparent 😅 )

AliSoftware reviewed (commented):
Haven't finished my review, will check back later, just submitting what I have so far

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Update Makefile and Ruby Gem after Rust CLI updates

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Better handle multiline fluent writing

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Extract cli tests from Makefile to external script

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Add GitHub commit status to pipeline

iangmaia committed:
Bump release version

iangmaia committed:
Update build automation setup

iangmaia committed:
Fix Rubocop violations and pwd call in cli-test.sh

iangmaia committed:
Remove `build` as dependency of `clit-test`

iangmaia committed:
Move Linux builds back to the `default` queue

AliSoftware commented on ruby/lib/fluent_tools/utils.rb:
```diff
@@ -0,0 +1,116 @@
+        "#{BINARY_NAME}-#{platform}.exe"
+      else
+        "#{BINARY_NAME}-#{platform}"
+      end
```
Let's use the helper you already have for that instead of duplicating their logic?
```suggestion
      extension = binary_extension(platform)
      "#{BINARY_NAME}-#{platform}#{extension}"
```

AliSoftware commented on ruby/lib/fluent_tools/converter.rb:
```diff
@@ -0,0 +1,128 @@
+
+    def development_context?
+      @project_root ||= find_project_root
+      !@project_root.nil?
```
`def project_root` already provides this, so let's just use it instead of duplicating code
```suggestion
      !project_root.nil?
```

AliSoftware commented on ruby/lib/fluent_tools/converter.rb:
```diff
@@ -0,0 +1,128 @@
+
+    def find_project_root
+      Pathname.new(__dir__).ascend do |dir|
+        return dir.to_s if File.exist?(File.join(dir, 'Cargo.toml'))
```
Given this is about finding the root of the project in the context of the Ruby lib / gem wrapper, I wonder if it wouldn't make more sense to check for `Gemfile` rather than `Cargo.toml`?

Ultimately I expect those to be in the same project root folder anyway, but just for consistency

AliSoftware commented on ruby/lib/fluent_tools/converter.rb:
```diff
@@ -0,0 +1,128 @@
+      nil
+    end
+
+    def validate_input_file(path)
```
I'd suggest to use `!` given it's validation that raises if it doesn't pass
```suggestion
    def validate_input_file!(path)
```

iangmaia replied:
Updated on 05d1a515d4134732c79de51d07d485d5f2e63227

AliSoftware commented on ruby/lib/fluent_tools/converter.rb:
💭 Nitpicky 2 cents:
I think during my first review of this (as a flat diff in GitHub without really realizing about the folder organization and this being in `ruby/lib`, the fact that this file was named `converter.rb` made me think this was supposed to contain the actual implementation for converting (as in, re-implementing the logic).
Of course that's not the case. But I wonder if picking a file and class name more like `invoke.rb` or `command_executor.rb` or `tool-invocation.rb` or something? (ok, maybe a bit of a mouthful…)

iangmaia replied:
I actually agree... And sometimes navigating the project I saw myself opening a Rust file instead of this one 😅 
Renamed it to `command_executor.rb` in dba76186737609048c6091928f740d7b2564b029.

AliSoftware commented on ruby/spec/fluent_tools_spec.rb:
```diff
@@ -0,0 +1,35 @@
+  describe '.fluent_to_android' do
+    it 'delegates to Converter' do
+      converter = instance_double(FluentTools::Converter)
+      allow(FluentTools::Converter).to receive(:new).and_return(converter)
```
Those could be moved in a `before do … end` block for all the tests in this spec to use instead of duplicating the mocking for each.

AliSoftware commented on ruby/Rakefile:
```diff
@@ -0,0 +1,186 @@
+  paths[:release_binary_name]
+end
+
+def find_or_create_release(client, repo, version)
```
Given all the mentions of "Release" in that `Rakefile` that could be misinterpreted to mean "Release build, as opposed to Dev/Debug build" instead of "GitHub Release", I think it'd be nice to make those task names more explicit:
```suggestion
def find_or_create_github_release(client, repo, version)
```

iangmaia replied:
Agreed 👍 Updated on a62d0355cd2b4de20105c99f0ee5b270e2f59c58.

AliSoftware commented on ruby/Rakefile:
```diff
@@ -0,0 +1,186 @@
+end
+
+desc 'Build release binary for platform and upload to GitHub releases (platform: PLATFORM, defaults to current platform)'
+task :release_binary, [:platform] do |_t, args|
```
Given all the mentions of "Release" in that Rakefile that could be misinterpreted to mean "Release build, as opposed to Dev/Debug build" instead of "GitHub Release", I think it'd be nice to make those task names more explicit.

Maybe something like `create_new_github_release`?

iangmaia replied:
Agreed 👍 to make the target more obvious, I'm calling it `build_and_create_github_release`. Updated on a62d0355cd2b4de20105c99f0ee5b270e2f59c58.

I've also removed on 7eaec6abf10345612aa75d5ada6f0fee364c67de a `release_all_platforms` that I implemented but ended up not using as it eventually proved impractical.

AliSoftware reviewed (approved):
Left some nitpicks and naming suggestions but nothing blocking, so approving to unblock.

iangmaia committed:
Use extension helper

Co-authored-by: Olivier Halligon <olivier@halligon.net>

iangmaia committed:
Remove redundant call to `find_project_root`

Co-authored-by: Olivier Halligon <olivier@halligon.net>

iangmaia committed:
Add ! to function to signal it can raise an error

iangmaia committed:
Rename `release_binary` Rake target to `build_and_create_github_release`

iangmaia committed:
Remove redundant (and impractical) `release_all_platforms` target

iangmaia committed:
Rename `FluentTools::Converter` to `FluentTools::CommandExecutor`

iangmaia committed:
Run RSpec together with Rubocop

iangmaia committed:
Update GitHub commit status to include tests

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Merge branch 'iangmaia/android-xml-support' into iangmaia/ruby-gem-support

iangmaia committed:
Fix cli-test script

iangmaia committed:
Merge branch 'trunk' into iangmaia/ruby-gem-support

iangmaia committed:
Enable Gem publishing

iangmaia committed:
Specify `android` agent for Linux builds
We need an agent with Ruby installed; using Docker would create complications as we also run Docker to compile

iangmaia committed:
Fix Gem push script to consider the `./ruby` dir

iangmaia committed:
Fix Rubocop offenses and RSpec tests
Fixed the RequireMFA Rule and tweaked Rubocop config for consistency

iangmaia committed:
Simplify Gem setup

iangmaia committed:
Fix Rubocop offense
